guppylang/checker/expr_checker.py

```diff
@@ -797,14 +798,37 @@ def check_type_against(
         # Finally, check that the instantiation respects the linearity requirements
         check_inst(act, inst, node)
 
-        return subst, inst
+        return node, subst, inst
 
     # Otherwise, we know that `act` has no unsolved type vars, so unification is trivial
     assert not act.unsolved_vars
     subst = unify(exp, act, {})
     if subst is None:
+        # Maybe we can implicitly coerce `act` to `exp`
+        if coerced := try_coerce_to(act, exp, node, ctx):
+            return coerced, {}, []
         raise GuppyTypeError(TypeMismatchError(node, exp, act, kind))
-    return subst, []
+    return node, subst, []
+
+
+def try_coerce_to(
+    act: Type, exp: Type, node: ast.expr, ctx: Context
+) -> ast.expr | None:
+    """Tries to implicitly coerce an expression to a different type.
+
+    Returns the coerced expression or `None` if the type cannot be implicitly coerced.
+    """
+    # Currently, we only support implicit coercions of numeric types
+    if not isinstance(act, NumericType) or not isinstance(exp, NumericType):
+        return None
+    # Ordering on `NumericType.Kind` defines the coercion relation
+    if act.kind < exp.kind:
+        f = ctx.globals.get_instance_func(act, f"__{exp.kind.name.lower()}__")
```

Comment:
so int -> float is ok but float -> int is not, correct?

Reply:
Exactly 👍